rejex.test.js

```diff
@@ -0,0 +1,170 @@
+  expect(regexMatch("h", new CharMatchExpr("h"))).toBe(true);
+  expect(regexMatch("he", new CharMatchExpr("h"))).toBe(false);
+});
+
+test('concat short', () => {
+  expect(regexMatch("he", new ConcatExpr(new CharMatchExpr("h"), new CharMatchExpr("e")))).toBe(true);
+});
+
+test('concat long', () => {
+  const expr =
+    new ConcatExpr(  // (hell)o
+      new ConcatExpr(  // (he)(ll)
+        new ConcatExpr(  // he
+          new CharMatchExpr("h"),
+          new CharMatchExpr("e")),
+        new ConcatExpr(  // ll
+          new CharMatchExpr("l"),
+          new CharMatchExpr("l"))),
+      new CharMatchExpr("o"))
+  expect(regexMatch("hello", expr)).toBe(true);
+});
+
+test('concat incomplete string', () => {
+  const expr =
+    new ConcatExpr(  // he
+      new CharMatchExpr("h"),
+      new CharMatchExpr("e"))
+  expect(regexMatch("h", expr)).toBe(false);
+});
+
+test('concat too much string', () => {
+  const expr =
+    new ConcatExpr(  // he
+      new CharMatchExpr("h"),
+      new CharMatchExpr("e"))
+  expect(regexMatch("hel", expr)).toBe(false);
+});
+
+test('concat with repetition', () => {
+  const expr =
+    new ConcatExpr(  // he*
+      new CharMatchExpr("h"),
+      new RepetitionExpr(  // e*
+        new CharMatchExpr("e")))
+  expect(regexMatch("h", expr)).toBe(true);
+  expect(regexMatch("he", expr)).toBe(true);
+  expect(regexMatch("hee", expr)).toBe(true);
+});
+
+test('alternation short', () => {
+  const expr =
+    new AlternationExpr(  // [he]
+      new CharMatchExpr("h"),
+      new CharMatchExpr("e"))
+  expect(regexMatch("", expr)).toBe(false);
+  expect(regexMatch("h", expr)).toBe(true);
+  expect(regexMatch("e", expr)).toBe(true);
+  expect(regexMatch("he", expr)).toBe(false);
+});
+
+test('alternation long', () => {
+  // [hello]
+  const expr =
+    new AlternationExpr(  // (hell)o
+      new AlternationExpr(  // (he)(ll)
+        new AlternationExpr(  // he
+          new CharMatchExpr("h"),
+          new CharMatchExpr("e")),
+        new AlternationExpr(
+          new CharMatchExpr("l"),
+          new CharMatchExpr("l"))),
+      new CharMatchExpr("o"))
+  expect(regexMatch("h", expr)).toBe(true);
+  expect(regexMatch("e", expr)).toBe(true);
+  expect(regexMatch("l", expr)).toBe(true);
+  expect(regexMatch("o", expr)).toBe(true);
+  expect(regexMatch("x", expr)).toBe(false);
+  expect(regexMatch("hello", expr)).toBe(false);
+  expect(regexMatch("", expr)).toBe(false);
+});
+
+test('repetition short', () => {
+  const expr =
+    new RepetitionExpr(  // h*
+      new CharMatchExpr("h"))
+  expect(regexMatch("", expr)).toBe(true);
+  expect(regexMatch("h", expr)).toBe(true);
+  expect(regexMatch("hh", expr)).toBe(true);
+  expect(regexMatch("hhh", expr)).toBe(true);
+  expect(regexMatch("hhhh", expr)).toBe(true);
+});
+
+test('repetition wrong char', () => {
+  // h*
+  const expr = new RepetitionExpr(new CharMatchExpr("h"))
+  expect(regexMatch("g", expr)).toBe(false);
+});
```

Comment:
I'm thinking maybe just combine this test with the one above it and add one more expect for a longer wrong repetition string like "ghh"?

Reply:
Done